### PR TITLE
Start hardening docker grading & optimize ci

### PR DESCRIPTION
**Pull request by fritterhoff**
As a first change for the issue #257

### CODE REVIEW

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -18,10 +18,18 @@ jobs:
...
     steps:
       - uses: actions/checkout@v2
+      
+      - uses: docker/setup-buildx-action@v1
+        id: buildx
+        with:
+          install: true
+
+      - name: Expose GitHub Runtime
+        uses: crazy-max/ghaction-github-runtime@v1
```
Comment (chrispyles) [inline .github/workflows/run-tests.yml:28]: What is this action used for?

File: Makefile [hunk 2]
```diff
@@ -14,6 +14,15 @@ docker-test:
 	docker build . -t otter-test -f test-Dockerfile --cache-from ucbdsinfra/otter-grader:latest
 	rm test-Dockerfile
 
+docker-ci-test:
+	cp -r Dockerfile test-Dockerfile
+	printf "\nADD . /home/otter-grader\nRUN pip install /home/otter-grader" >> test-Dockerfile
+	printf "\nRUN cd /tmp && curl -sSL -O https://download.docker.com/linux/static/stable/x86_64/docker-${DOCKER_VERSION}.tgz && tar zxf docker-${DOCKER_VERSION}.tgz && mv ./docker/docker /usr/local/bin && chmod +x /usr/local/bin/docker && rm -rf /tmp/*" >> test-Dockerfile
+	printf "\nCOPY --from=docker/buildx-bin /buildx /usr/libexec/docker/cli-plugins/docker-buildx" >> test-Dockerfile
+	printf "\nENV PYTHONUNBUFFERED 1" >> test-Dockerfile
+	docker buildx build . --load -t otter-test -f test-Dockerfile --cache-from=type=gha --cache-to=type=gha,mode=max
```
Comment (chrispyles) [inline Makefile:23]: Can we out the first 5 lines of this target into a separate one that `docker-test` and `docker-ci-test` rely on? e.g.

```
docker-test-precursor:
	# those lines

docker-test: docker-test-precursor
	# etc.

docker-ci-test: docker-test-precursor
	# etc.
```

File: otter/grade/containers.py [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
-import pkg_resources
 import shutil
 import tempfile
-
 from concurrent.futures import ThreadPoolExecutor, wait
+from typing import Optional
+
+import pandas as pd
+import pkg_resources
+from python_on_whales import docker
```
Comment (chrispyles) [inline otter/grade/containers.py:12]: do you mind re-ordering these imports to match the style? sorry for the nit-picking but I'm trying to slowly-but-surely fix a but of mismatched styles in this repo and would like all new contributes not to introduce new style issues.

order should be

* `import ...< as ...>` statements in alphabetical order
* &lt;empty line&gt;
* `from ... import ...` statements _not from within the `otter` package_ in alphabetical order
* &lt;empty line&gt;
* `from ... import ...` statements importing from within the `otter` package.

Thanks!

File: test/test_grade.py [hunk 4]
```diff
@@ -73,11 +74,48 @@ def test_docker(self):
...
+                containers=5,
+                image="otter-test",
+                timeout=40,
+                debug=True,
+                verbose=True,
+            )
+        self.assertEqual(str(e.exception),"Executing 'test/test-grade/timeout/20s.ipynb' in docker container failed! Exit code: 137")
+
+    def test_network(self):
+        grade(
```
Comment (chrispyles) [inline test/test_grade.py:93]: Can you add a short docstring describing what this method tests?

File: test/test_grade.py [hunk 5]
```diff
@@ -73,11 +74,48 @@ def test_docker(self):
         # assert that it didn't fail, it will fail if it is not installed
         self.assertEqual(len(inspect.stderr), 0, inspect.stderr.decode("utf-8"))
 
+
+    def test_timeout(self):
+        with self.assertRaises(Exception) as e:
```
Comment (chrispyles) [inline test/test_grade.py:79]: Can you add a short docstring describing what this method tests?

File: test/test_grade.py [hunk 6]
```diff
@@ -124,12 +162,10 @@ def test_notebooks_with_pdfs(self):
...
         self.assertEqual(sorted(dir1_contents), sorted(dir2_contents), f"'{TEST_FILES_PATH}notebooks/' and 'test/submission_pdfs' have different contents")
 
-
-
+    def tearDown(self) -> None:
         # remove the extra output
-        cleanup_command = ["rm", "-rf", "test/final_grades.csv", "test/submission_pdfs", "test/final_grades.csv", TEST_FILES_PATH + "autograder.zip"]
+        cleanup_command = ["rm", "-rf", "test/final_grades.csv", "test/submission_pdfs", "test/final_grades.csv"]
         cleanup = subprocess.run(cleanup_command, stdout=PIPE, stderr=PIPE)
-        self.assertEqual(len(cleanup.stderr), 0, cleanup.stderr.decode("utf-8"))
```
Comment (chrispyles) [inline test/test_grade.py:132]: How come you removed this line? I think with `PIPE` the cleanup process won't reveal any errors if we don't perform this check.

**Review by chrispyles — CHANGES_REQUESTED** [review]
Left a few comments and questions. Thanks for these changes @fritterhoff!

File: otter/grade/containers.py (hunk 3 above)
Comment (fritterhoff) [reply otter/grade/containers.py:12]: No Problem. That's the fault of pycharm. I'll try to change the settings (:

File: test/test_grade.py (hunk 6 above)
Comment (fritterhoff) [reply test/test_grade.py:132]: Well that was a mistake.

File: .github/workflows/run-tests.yml (hunk 1 above)
Comment (fritterhoff) [reply .github/workflows/run-tests.yml:28]: Friends gave me the hint how to better utilize buildkit in github actions. You can use buildx in combination with github actions caching (`--cache-from=type=gha` and `--cache-to=type=gha`) 
Therefore you must enable buildkit and expose the runtime so `docker buildx build` can access the variables inside the run step.

**Review by chrispyles — APPROVED** [review]
Thanks for resolving those. LGTM otherwise so feel free to mark this PR as ready when you are.

**Pull request merged**